plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs

```diff
@@ -0,0 +1,188 @@
+-}
+module PlutusIR.Transform.LetFloatIn (floatTerm) where
+
+import PlutusCore qualified as PLC
+import PlutusCore.Builtin qualified as PLC
+import PlutusCore.Name qualified as PLC
+import PlutusIR
+import PlutusIR.Purity (isPure)
+import PlutusIR.Transform.Rename ()
+
+import Control.Lens hiding (Strict)
+import Control.Monad ((<=<))
+import Data.Set (Set)
+import Data.Set qualified as Set
+
+type Uniques = Set PLC.TermUnique
+
+-- | Float bindings in the given `Term` inwards.
+floatTerm ::
+    forall m tyname name uni fun a.
+    ( PLC.HasUnique name PLC.TermUnique
+    , PLC.HasUnique tyname PLC.TypeUnique
+    , PLC.ToBuiltinMeaning uni fun
+    , PLC.MonadQuote m
+    ) =>
+    PLC.BuiltinVersion fun ->
+    Term tyname name uni fun a ->
+    m (Term tyname name uni fun a)
+floatTerm ver = pure . fmap fst . go <=< PLC.rename
+  where
+    -- | Float bindings in the given `Term` inwards, and calculate the set of
+    -- variable `Unique`s in the result `Term`.
+    go ::
+        Term tyname name uni fun a ->
+        Term tyname name uni fun (a, Uniques)
+    go t = case t of
+        Apply a fun0 arg0 ->
+            let fun = go fun0
+                arg = go arg0
+                us = termUniqs fun `Set.union` termUniqs arg
+             in Apply (a, us) fun arg
+        LamAbs a n ty body0 ->
+            let body = go body0
+             in LamAbs (a, termUniqs body) n (noUniq ty) body
+        TyAbs a n k body0 ->
+            let body = go body0
+             in TyAbs (a, termUniqs body) n (noUniq k) body
+        TyInst a body0 ty ->
+            let body = go body0
+             in TyInst (a, termUniqs body) body (noUniq ty)
+        IWrap a ty1 ty2 body0 ->
+            let body = go body0
+             in IWrap (a, termUniqs body) (noUniq ty1) (noUniq ty2) body
+        Unwrap a body0 ->
+            let body = go body0
+             in Unwrap (a, termUniqs body) body
+        Let a NonRec bs0 body0 ->
+            let bs = go' <$> bs0
+                body = go body0
+             in -- The bindings in `bs` should be processed from right to left, since
+                -- a binding may depend on another binding to its left.
+                foldr (floatAlg ver a) body bs
+        Let a Rec bs0 body0 ->
+            -- Currently we don't move recursive bindings, so we simply descend into the body.
+            let bs = go' <$> bs0
+                body = go body0
+                us = Set.union (termUniqs body) (foldMap bindingUniqs bs)
+             in Let (a, us) Rec bs body
+        Var a n -> Var (a, Set.singleton (n ^. PLC.unique)) n
+        Constant{} -> (,mempty) <$> t
+        Builtin{} -> (,mempty) <$> t
+        Error{} -> (,mempty) <$> t
+
+    -- | Float bindings in the given `Binding` inwards, and calculate the set of
+    -- variable `Unique`s in the result `Binding`.
+    go' ::
+        Binding tyname name uni fun a ->
+        Binding tyname name uni fun (a, Uniques)
+    go' = \case
```

Comment:
Again, this would be easy to do with a traversal if you weren't using the annotation field